src/UserSpaceInstrumentation/TrampolineTest.cpp

```diff
@@ -0,0 +1,249 @@
+  AddressRange b4 = {1, 9};
+  EXPECT_TRUE(DoAddressRangesOverlap(a, b4));
+  AddressRange b5 = {4, 5};
+  EXPECT_TRUE(DoAddressRangesOverlap(a, b5));
+  AddressRange b6 = {4, 9};
+  EXPECT_TRUE(DoAddressRangesOverlap(a, b6));
+  AddressRange b7 = {7, 9};
+  EXPECT_FALSE(DoAddressRangesOverlap(a, b7));
+  AddressRange b8 = {8, 9};
+  EXPECT_FALSE(DoAddressRangesOverlap(a, b8));
+}
+
+TEST(TrampolineTest, LowestIntersectingAddressRange) {
+  const std::vector<AddressRange> kAllRanges = {{0, 5}, {20, 30}, {40, 60}};
+
+  EXPECT_FALSE(LowestIntersectingAddressRange({}, {0, 60}).has_value());
+
+  EXPECT_EQ(0, LowestIntersectingAddressRange(kAllRanges, {1, 2}));
+  EXPECT_EQ(1, LowestIntersectingAddressRange(kAllRanges, {21, 22}));
+  EXPECT_EQ(2, LowestIntersectingAddressRange(kAllRanges, {51, 52}));
+
+  EXPECT_EQ(0, LowestIntersectingAddressRange(kAllRanges, {3, 6}));
+  EXPECT_EQ(1, LowestIntersectingAddressRange(kAllRanges, {19, 22}));
+  EXPECT_EQ(2, LowestIntersectingAddressRange(kAllRanges, {30, 52}));
+
+  EXPECT_EQ(0, LowestIntersectingAddressRange(kAllRanges, {4, 72}));
+  EXPECT_EQ(1, LowestIntersectingAddressRange(kAllRanges, {29, 52}));
+  EXPECT_EQ(2, LowestIntersectingAddressRange(kAllRanges, {59, 72}));
+
+  EXPECT_FALSE(LowestIntersectingAddressRange(kAllRanges, {5, 20}).has_value());
+  EXPECT_FALSE(LowestIntersectingAddressRange(kAllRanges, {30, 40}).has_value());
+  EXPECT_FALSE(LowestIntersectingAddressRange(kAllRanges, {60, 80}).has_value());
+}
+
+TEST(TrampolineTest, HighestIntersectingAddressRange) {
+  std::vector<AddressRange> kAllRanges = {{0, 5}, {20, 30}, {40, 60}};
+
+  EXPECT_FALSE(HighestIntersectingAddressRange({}, {0, 60}).has_value());
+
+  EXPECT_EQ(0, HighestIntersectingAddressRange(kAllRanges, {1, 2}));
+  EXPECT_EQ(1, HighestIntersectingAddressRange(kAllRanges, {21, 22}));
+  EXPECT_EQ(2, HighestIntersectingAddressRange(kAllRanges, {51, 52}));
+
+  EXPECT_EQ(0, HighestIntersectingAddressRange(kAllRanges, {3, 6}));
+  EXPECT_EQ(1, HighestIntersectingAddressRange(kAllRanges, {19, 22}));
+  EXPECT_EQ(2, HighestIntersectingAddressRange(kAllRanges, {30, 52}));
+
+  EXPECT_EQ(2, HighestIntersectingAddressRange(kAllRanges, {4, 72}));
+  EXPECT_EQ(2, HighestIntersectingAddressRange(kAllRanges, {29, 52}));
+  EXPECT_EQ(2, HighestIntersectingAddressRange(kAllRanges, {59, 72}));
+
+  EXPECT_FALSE(HighestIntersectingAddressRange(kAllRanges, {5, 20}).has_value());
+  EXPECT_FALSE(HighestIntersectingAddressRange(kAllRanges, {30, 40}).has_value());
+  EXPECT_FALSE(HighestIntersectingAddressRange(kAllRanges, {60, 80}).has_value());
+}
+
+TEST(TrampolineTest, FindAddressRangeForTrampoline) {
+  constexpr uint64_t k64Kb = 0x10000;
+  constexpr uint64_t kOneMb = 0x100000;
+  constexpr uint64_t k256Mb = 0x10000000;
+  constexpr uint64_t kOneGb = 0x40000000;
+
+  // Trivial placement to the left.
+  const std::vector<AddressRange> kTakenRanges1 = {
+      {0, k64Kb}, {kOneGb, 2 * kOneGb}, {3 * kOneGb, 4 * kOneGb}};
```

Comment:
Maybe add a test for ranges which do not have the very first one start with 0 address?

Reply:
Added EXPECT_DEATH test.